### ISSUE TITLE
pull 406 breaks Activate/Delete links when subdomains are used

### ISSUE THREAD

**Issue by arno01** [posativ/isso#420]
Hi,

It appears that https://github.com/posativ/isso/pull/406 breaks Activate/Delete links when subdomains are used.

In my case I have isso at: comments.mysite.com
So the links coming in email are invalid:

```
Delete comment: https://mysite.com//id/123/delete/MTI...redacted
Activate comment: https://mysite.com//id/123/activate/MTI...redacted
```

> Note also there is a double slash ``//`` which should not be there.

Applying this (pretty much a revert), fixes that, so it is working well as before:

```
diff --git a/isso/ext/notifications.py b/isso/ext/notifications.py
index 2b4a81c..2900984 100644
--- a/isso/ext/notifications.py
+++ b/isso/ext/notifications.py
@@ -115,7 +115,7 @@ class SMTP(object):
                  (local("origin") + thread["uri"] + "#isso-%i" % comment["id"]))
         rv.write("\n")
 
-        uri = self.general_host + "/id/%i" % comment["id"]
+        uri = local("host") + "/id/%i" % comment["id"]
         key = self.isso.sign(comment["id"])
 
         rv.write("---\n")
```

Good links for subdomain configuration should be:

```
Delete comment: https://comments.mysite.com/id/123/delete/MTI...redacted
Activate comment: https://comments.mysite.com/id/123/activate/MTI...redacted
```

----

We still have to fix isso when its endpoint is a directory instead of a subdomain https://github.com/posativ/isso/issues/401

**Comment by pellenilsson** [comment]
This hit me as well. I run Isso at a different port than the web server. The setting `isso.conf.get("general", "host")` is supposed to contain the URL of the website, not the Isso server. The setting `isso.conf.get("server", "listen")` would work in some cases, but probably not in all cases since it can be a Unix domain socket (`unix:///tmp/isso.sock`) or a local address (`http://localhost:1234/`).

I don't have the full picture of all deployment methods, so I can't really suggest a fix unfortunately. The old code (`local("host")`) only worked for uWSGI?

**Comment by pellenilsson** [comment]
Hmm, in the general case I guess it's impossible for Isso to auto-detect the public URL that it can be accessed with from the outside world, since part of the configuration may be in another application (e.g. Apache or nginx). So maybe we have to introduce a new setting for that (`server.public_url` or something). If that setting is missing, it could perhaps try auto detection the old way (`local("host")`). Sounds good?

ping @ycpei

**Comment by blatinier** [comment]
@pellenilsson I agree on the solution including a new setting and fallback to `local("host")`. 
If you include it in you PR #443 it could be a great package :)
I'll have to create some time to test that because I definetly want it merged.

**Comment by ycpei** [comment]
Thanks for pinging, and sorry for the problem and for not seeing this thread earlier - I was only made aware of this thread because of the mention. Off-topic question: is there a way on github to be notified of references to PRs one is involved in?

It seems the overall situation is more complicated than I thought, and I'm afraid I can't make any comments / suggestions without spending some time understanding isso.

**Comment by pellenilsson** [comment]
No problem @ycpei. I will try to implement my proposal asap. It may take a little while due to exceptionally good weather here in Stockholm.

**Comment by ycpei** [comment]
@pellenilsson Thanks. I am also in Stockholm and agree the weather is overall amazing, but the sun could do with a bit less glare :)

**Comment by pellenilsson** [comment]
Incorporated my suggestion into #443 now. Tested it both with and without the new setting present.

**Comment by craigphicks** [comment]
This may or may not be relevant:
With Issp served by nginx on either of these sites:

    https:\\sameasblog.com/isso
   https:\\notsameasblog.com/isso

the admin pages were not working.

Nginx setup:

    location /isso {
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Script-Name /isso;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://localhost:1234;
    }

Note the critical line `proxy_set_header X-Script-Name /isso;`

I had to change the isso code to make it work.  Mostly it was fixed by using the variable `local.host` where it should be used so that it shows up in *login.html* and *admin.html*.  Even if a config file solution is available, `local.host` should be correctly used to enable the default when possible.

The change to use `local.host` is in [this pull request](https://github.com/posativ/isso/pull/458).
